src/crypto/parse.rs

```diff
@@ -14,46 +20,89 @@ pub enum ParseHexError {
     MalformedHex { backtrace: Backtrace },
 }
 
-trait ParseHex<To> {
-    fn parse_hex(&self) -> Result<To, ParseHexError>;
+trait ParseHex {
+    fn parse_hex(&self) -> Result<Vec<u8>, ParseHexError>;
 }
 
-macro_rules! implement {
-    ($($size:expr), *) => ($(
-        impl ParseHex<[u8; $size]> for str {
-            fn parse_hex(&self) -> Result<[u8; $size], ParseHexError> {
-                ensure!(self.len() == (2 * $size), UnexpectedSize{});
-                    let mut parsed = [u8::default(); $size];
-                    for index in 0..$size {
-                        parsed[index] = u8::from_str_radix(&self[(2 * index)..(2 * (index + 1))], 16)
-                            .map_err(|_| MalformedHex.build())?;
-                    }
-
-                    Ok(parsed)
+impl ParseHex for &str {
+    fn parse_hex(&self) -> Result<Vec<u8>, ParseHexError> {
+        let mut parsed = Vec::new();
+
+        ensure!(self.len() % 2 == 0, UnexpectedSize);
+
+        for index in (0..self.len()).step_by(2) {
+            let value = u8::from_str_radix(&self[index..(index + 2)], 16)
+                .map_err(|_| MalformedHex.build())?;
+
+            parsed.push(value);
+        }
```

Comment:
mostly a comment for me to understand better/play with rust.
is it a better way to implement it or not?

```suggestion
        ensure!(self.len() % 2 == 0, UnexpectedSize);

        let parsed = (0..self.len())
            .step_by(2) ‣impl Iterator<Item = usize>
            .map(|index| u8::from_str_radix(&self[index..(index + 2)], 16))
            .collect::<Result<Vec<_>, _>>()
            .map_err(|_| MalformedHex.build())?;
```

Reply:
Both are equivalent though I prefer for loops when using indices

Reply:
all right :+1:
(btw, I can't "resolve discussion" myself, so you'll have to close each you think was fix)